pkg/xds/translate.go

```diff
@@ -0,0 +1,302 @@
+	out := route.RouteMatch{PathSpecifier: &route.RouteMatch_Prefix{Prefix: "/"}}
+	if in == nil {
+		return out
+	}
+
+	for name, stringMatch := range in.Headers {
+		matcher := TranslateHeaderMatcher(name, stringMatch)
+		out.Headers = append(out.Headers, &matcher)
+	}
+
+	// guarantee ordering of headers
+	sort.Slice(out.Headers, func(i, j int) bool {
+		if out.Headers[i].Name == out.Headers[j].Name {
+			return out.Headers[i].Value < out.Headers[j].Value
+		}
+		return out.Headers[i].Name < out.Headers[j].Name
+	})
+
+	if in.Uri != nil {
+		switch m := in.Uri.MatchType.(type) {
+		case *routingv2.StringMatch_Exact:
+			out.PathSpecifier = &route.RouteMatch_Path{Path: m.Exact}
+		case *routingv2.StringMatch_Prefix:
+			out.PathSpecifier = &route.RouteMatch_Prefix{Prefix: m.Prefix}
+		case *routingv2.StringMatch_Regex:
+			out.PathSpecifier = &route.RouteMatch_Regex{Regex: m.Regex}
+		}
+	}
+
+	if in.Method != nil {
+		matcher := TranslateHeaderMatcher(HeaderMethod, in.Method)
+		out.Headers = append(out.Headers, &matcher)
+	}
+
+	if in.Authority != nil {
+		matcher := TranslateHeaderMatcher(HeaderAuthority, in.Authority)
+		out.Headers = append(out.Headers, &matcher)
+	}
+
+	if in.Scheme != nil {
+		matcher := TranslateHeaderMatcher(HeaderScheme, in.Scheme)
+		out.Headers = append(out.Headers, &matcher)
+	}
+
+	// TODO: match.DestinationPorts
+
+	return out
+}
+
+// TranslateHeaderMatcher translates to HeaderMatcher
+func TranslateHeaderMatcher(name string, in *routingv2.StringMatch) route.HeaderMatcher {
+	out := route.HeaderMatcher{
+		Name: name,
+	}
+
+	switch m := in.MatchType.(type) {
+	case *routingv2.StringMatch_Exact:
+		out.Value = m.Exact
+	case *routingv2.StringMatch_Prefix:
+		// Envoy regex grammar is ECMA-262 (http://en.cppreference.com/w/cpp/regex/ecmascript)
+		// Golang has a slightly different regex grammar
+		out.Value = fmt.Sprintf("^%s.*", regexp.QuoteMeta(m.Prefix))
+		out.Regex = &types.BoolValue{Value: true}
+	case *routingv2.StringMatch_Regex:
+		out.Value = m.Regex
+		out.Regex = &types.BoolValue{Value: true}
+	}
+
+	return out
+}
+
+// TranslateRetryPolicy translates retry policy
+func TranslateRetryPolicy(in *routingv2.HTTPRetry) *route.RouteAction_RetryPolicy {
+	if in != nil && in.Attempts > 0 {
+		return &route.RouteAction_RetryPolicy{
+			NumRetries:    &types.UInt32Value{Value: uint32(in.GetAttempts())},
+			RetryOn:       "5xx,connect-failure,refused-stream",
+			PerTryTimeout: TranslateTime(in.PerTryTimeout),
+		}
+	}
+	return nil
+}
+
+// TranslateCORSPolicy translates CORS policy
+func TranslateCORSPolicy(in *routingv2.CorsPolicy) *route.CorsPolicy {
+	if in == nil {
+		return nil
+	}
+
+	out := route.CorsPolicy{
+		AllowOrigin: in.AllowOrigin,
+		Enabled:     &types.BoolValue{Value: true},
+	}
+	if in.AllowCredentials != nil {
+		out.AllowCredentials = TranslateBool(in.AllowCredentials)
+	}
+	if len(in.AllowHeaders) > 0 {
+		out.AllowHeaders = strings.Join(in.AllowHeaders, ",")
```

Comment:
nit: don't need `len(in.AllowHeaders) > 0` guard here. `strings.Join` should return "" if the input is empty

Reply:
thanks, will fix it up here -- and probably can be fixed up in v1 version.

Reply:
Yes, it is a problem in the original code.

Reply:
fixed in a follow-up